client/allochealth/tracker.go

```diff
@@ -503,8 +517,24 @@ OUTER:
 		// Detect if all the checks are passing
 		passed := true
 
+		// interpolate services to replace runtime variables
+		consulServices := t.tg.ConsulServices()
+		interpolatedServices := make([]*structs.Service, len(consulServices))
+		for i, service := range consulServices {
+			env := t.taskEnvs[service.TaskName]
+			if env == nil {
+				// This is not expected to happen, but guard against a nil
+				// task environment by using the group environment since it has
+				// most of the same values.
+				t.logger.Warn("task environment not found, using group level environment for variable interpolation",
+					"alloc_id", t.alloc.ID, "task", service.TaskName)
+				env = t.taskEnvs[""]
```

Comment:
we [should] already guard against this in Validate/Canonicalize; instead of trying to make things _maybe_ work in an unexpected situation I think we should just do nothing, e.g.

```
if env != nil {
 // assign
} else {
  logger.Error("you found a bug!")
}
```

Reply:
Hum...good point. Changed to this in 925d34798